tt_metal/hw/firmware/src/brisc.cc

```diff
@@ -354,43 +357,69 @@ int main() {
     // Wait for ncrisc to halt
     wait_for_ncrisc_to_halt();
 
-    mailboxes->launch.go.run = RUN_MSG_DONE;
+    mailboxes->go_message.signal = RUN_MSG_DONE;
 
     while (1) {
         init_sync_registers();
         reset_ncrisc_with_iram();
 
         WAYPOINT("GW");
-        while (mailboxes->launch.go.run != RUN_MSG_GO);
+
+        while (mailboxes->go_message.signal != RUN_MSG_GO) {
+            // While the go signal for kernel execution is not sent, check if the worker was signalled
+            // to reset its launch message read pointer.
+            if (mailboxes->go_message.signal == RUN_MSG_RESET_READ_PTR) {
+                // Set the rd_ptr on workers to specified value
+                mailboxes->launch_msg_rd_ptr = 0;
+                uint64_t dispatch_addr =
+                    NOC_XY_ADDR(NOC_X(mailboxes->go_message.master_x),
+                    NOC_Y(mailboxes->go_message.master_y), DISPATCH_MESSAGE_ADDR);
+                mailboxes->go_message.signal = RUN_MSG_DONE;
+                // Notify dispatcher that this has been done
+                DEBUG_SANITIZE_NOC_ADDR(noc_index, dispatch_addr, 4);
+                noc_fast_atomic_increment(
+                    noc_index,
+                    NCRISC_AT_CMD_BUF,
+                    dispatch_addr,
+                    NOC_UNICAST_WRITE_VC,
+                    1,
+                    31 /*wrap*/,
+                    false /*linked*/);
+            }
+        }
+
+        uint32_t launch_msg_rd_ptr = mailboxes->launch_msg_rd_ptr;
         WAYPOINT("GD");
 
         {
+            // Only include this iteration in the device profile if the launch message is valid. This is because all workers get a go signal regardless of whether
+            // they're running a kernel or not. We don't want to profile "invalid" iterations.
```

Comment:
reading through L1 is expensive, create a pointer in a register for mailboxes->launch[launch_msg_rd_ptr] and access fields below through that pointer to be 100% sure the compiler behaves (any store through a pointer could cause a reload).

Reply:
Latest commit uses a pointer to remove the amount of memory indirection/L1 interaction. Take a look at erisc.cc if you like comedy. Was doing something much worse there.